### ISSUE TITLE
bitcoin.op_push(255) & bitcoin.push_script('00'*255) use '4dff00' instead of '4cff'

### ISSUE THREAD

**Issue by TinosNitso** [Electron-Cash/Electron-Cash#2376]
Lines 203 & 205 from [bitcoin.py](https://github.com/Electron-Cash/Electron-Cash/blob/master/electroncash/bitcoin.py) could use correction. Line 203 could become 'elif i<0x100:' or 'elif i<=0xff:' instead of 'elif i<0xff:'. Similarly for line 205. Until then plugin devs can write their own push_script function to handle 255B. The error can be produced in the **console** by entering `bitcoin.op_push(255)` or `bitcoin.push_script('00'*255)`.

Another issue is that when a `window.broadcast_transaction` fails for this reason, we get the same error msg as for an OpCode failure, which can lead smart contract devs on a big goose chase! The last thing I suspected was the 4dff00 versus 4cff issue. Pushing the raw TX using **Blockchair** gives a more detailed error message (needlessly large data push is what it said).

255B data-pushes can occur for scriptCode that size, or a Preimage with scriptCode 99B long. The case which got me was pushing a raw [TX](https://www.blockchain.com/bch/tx/c53298afa027f6eef5804bdaf8078c5887f5983f0ad4536d49b0445249a3cee1) whose size was 255B (uncompressed PubKey with ECDSA sig paying a P2SH output & P2PKH output). Unfortunately I blamed the official node software claiming they had a faulty OP_SPLIT or OP_BIN2NUM (but only for 128≤n<256)! (oops)

**Comment by cculianu** [comment]
I saw your issue on BCHN , and the wild goose chase that entailed. Let me look at the code and get back to you. You are probably right it's probably off-by-1  -- let me check.

As for the error message -- let me see what bitcoind returns in that case...

**Comment by cculianu** [comment]
Yep. There is indeed a bug.  It's insane that this bug has been here for many years.  The reason nobody noticed is because the most common usage of `push_script` in the app (which ends up calling `op_push`) is to create a p2sh multisig redeemscript.  And it's impossible to make a redeemscript that is exactly a 255 byte push.  So nobody noticed.

Anyway, I just pushed a commit to fix this: 710ec6436a002d52f0a77f8a346eec5a591a870f

May I ask what you are using Electron Cash for that you caught this bug? You piqued my curiosity...

**Comment by TinosNitso** [comment]
Thanks for the fix! I have this idea to create a _preturn_... vanity address that always returns whatever coins are sent to it, and can't do anything else. To be fully valid (always work & immune to malleability) & simple & efficient requires the new introspection OpCodes, but the one I've created is like a preliminary version ([AutoCove-Plugin](https://github.com/TinosNitso/AutoCove-Plugin)). It requires pushing the sender's raw TX into the scriptSig, and that can be exactly 255B. That data-push can prove who sent the money.

Fully auto covenants actually require a full node, though, to check all the OP_RETURN outputs in the whole blockchain. So someone's rent can be forwarded automatically by some node when it's due. EC could be for the renter in that case, but maybe not the node that actually broadcasts payments to the landlord.

The plugin can currently decode script hex into a more readable form. I [posted](https://www.reddit.com/r/btc/comments/rg0cid/decoding_p2sh_scripts_using_ec_plugin_autocove/) a short video on Reddit with some examples. The next version will be able to decode a raw TX into all the P2SH Scripts it used (user can select from the combo-box). It's mostly out of curiosity, so users can check all the scripts they use. A nice script "decoder" could conceivably help EC, if some scripts become popular.

**Comment by cculianu** [comment]
Ok cool! That sounds awesome and fun!  I'll check out the video and your reddit post.

Well I just cleaned up the `Script.push_data` and `bitcoin.py` `push_script` APIs to be 100% correct always. They were not respecting BIP62 "minimal push" rules. See: https://en.bitcoin.it/wiki/BIP_0062#Push_operators.

Thanks for the bug report. This made me realize that the script pushing in Electron Cash was broken, first with that off-by-1 bug and also with respect to BIP62 "minimal pushes".  I pushed a bunch of commits to master now to fix the situation.

Hopefully the plugin will be released after the next release version of electron cash...
